src/saving/ISaveUpgradeStep.cs

```diff
@@ -162,6 +163,174 @@ private void UpgradeBehaviouralValues(JProperty property, JEnumerable<JProperty>
+        ///         ...
+        ///       }
+        ///       ...
+        ///   </para>
+        /// </remarks>
+        private void UpgradePatchesVolume(JObject container, JEnumerable<JProperty> children)
+        {
+            if (children.First(p => p.Name == "Depth").Value is JArray depth)
+            {
+                if (depth[1].Value<int?>() == null || depth[0].Value<int?>() == null)
+                {
+                    throw new ArgumentException("Depth array does not start with two integers!");
+                }
+
+                var depthDifference = depth[1].Value<int>() - depth[0].Value<int>();
+
+                // Assume cubic patches for upgrade
+                container.Add("Volume", (long)Math.Pow(depthDifference, 3));
+            }
+            else
+            {
+                throw new ArgumentException("Depth value is not an array!");
+            }
+        }
+
+        /// <summary>
+        ///   Upgrades registered effect by adding gas production
+        /// </summary>
+        /// <param name="property">the property registering the effectes</param>
+        /// <param name="gasProductionEffectName">the pathname of the effect to add</param>
+        /// <remarks>
+        ///   <para>
+        ///     Changes a JSON like
+        ///       "effects":[{
+        ///         "$type": "GlucoseReductionEffect, Thrive",
+        ///         "targetWorld": {
+        ///           "$ref": "2"
+        ///         }
+        ///       }]
+        ///     to
+        ///       "effects":[{
+        ///         "$type": "GlucoseReductionEffect, Thrive",
+        ///         "targetWorld": {
+        ///           "$ref": "2"
+        ///         }
+        ///       },
+        ///       {
+        ///         "$type": "GasProductionEffect, Thrive",
+        ///         "targetWorld": {
+        ///           "$ref": "2"
+        ///         }
+        ///       }]
+        ///   </para>
+        /// </remarks>
+        private void UpgradeGasProductionEffect(JProperty property, string gasProductionEffectName)
+        {
+            if (property.Value is JArray effectsArray)
+            {
+                if (effectsArray.Count == 0)
+                {
+                    throw new ArgumentException("Property " + property.Name +
+                        " did not have registered tokens in array.");
+                }
+
+                string targetWorldRef = effectsArray[0]["targetWorld"]?["$ref"]?.ToString() ??
+                    throw new JsonException("Missing field from the first element of effect array: " +
+                        "targetWorld or its $ref!");
+
+                var typeTokenText = "'$type': '" + gasProductionEffectName + "'";
+                var targetWorldTokenText = "'targetWorld': { '$ref': '" + targetWorldRef + "' }";
+                var effectTokenText = "{" + typeTokenText + "," + targetWorldTokenText + "}";
+                var effectToken = JToken.Parse(effectTokenText);
```

Comment:
Wouldn't it be better to actually build the JObject properly instead of doing a bunch of string concatenation?
(I've seen some r/programminghorrorposts about manually created JSON strings...)